pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -525,6 +526,9 @@ public boolean index(GenericRow row, @Nullable RowMetadata rowMetadata)
       }
     }
 
+    // NOTE: we must do this before we index a single column to avoid partially indexing the row
```

Comment:
Can we add link to the current hard limit?

Reply:
@jackluo923 can you add a unit test to cover this function/scenario?

Reply:
@deemoliu After inspecting the code and writing a unit test, it seems the problem is not reproducible in a unit test because Java is running in `non-production` mode (i.e. `java -ea xxx`) with assertions enabled. This validation check is pretty much equivalent to the assertion found in `org.apache.pinot.segment.local.realtime.impl.forward.FixedByteMVMutableForwardIndex#updateHeader`, except we now throw an exception at run-time.:

```
  private int updateHeader(int row, int numValues) {
    assert (numValues <= _maxNumberOfMultiValuesPerRow);
    int newStartIndex = _prevRowStartIndex + _prevRowLength;
    if (newStartIndex + numValues > _currentCapacity) {
      addDataBuffer(_incrementalCapacity);
      _prevRowStartIndex = 0;
      _prevRowLength = 0;
      newStartIndex = 0;
    }
    writeIntoHeader(row, _dataWriters.size() - 1, newStartIndex, numValues);
    _prevRowStartIndex = newStartIndex;
    _prevRowLength = numValues;
    return newStartIndex;
  }
```

In production-mode, the assert is an no-op and if a value inserted into the MV forward index is greater than the maximum, we get inconsistencies later on in the code. In `validateLengthOfMVColumns`, we use a more relaxed check which validate against `_incrementalCapacity`. FixedByteMVMutableForwardIndex derives the maximum # of values per entry using `_maxNumberOfMultiValuesPerRow`, but `_incrementalCapacity` is the real maximum. See code in `org.apache.pinot.segment.local.realtime.impl.forward.FixedByteMVMutableForwardIndex#FixedByteMVMutableForwardIndex` for more detail.

Reply:
@chenboat  
The current maximum is 1000 (default) entries for MV forward indexes because the code paths which invokes the constructor almost always use the default value. See `org.apache.pinot.segment.local.segment.index.forward.ForwardIndexType#MAX_MULTI_VALUES_PER_ROW` for the default value of 1000. See `org.apache.pinot.segment.local.segment.index.forward.ForwardIndexType#createMutableIndex` for the 2 places where MV mutable forward indexes is created with this 1000 max value limit.